### ISSUE TITLE
WebRTC-SVC (Scalable Video Coding)

### ISSUE THREAD

**Issue by aboba** [w3ctag/design-reviews#396]
こんにちはTAG!

I'm requesting a TAG review of:

  - Name: WebRTC-SVC
  - Specification URL: https://w3c.github.io/webrtc-svc/
  - Explainer (containing user needs and example code)¹: https://github.com/w3c/webrtc-svc/blob/master/explainer.md
  - GitHub issues (if you prefer feedback filed there): https://github.com/w3c/webrtc-svc/issues
  - Tests: No WPT tests yet
  - Primary contacts (and their relationship to the specification): [@pthatcherg, @aboba]

Further details:

  - Relevant time constraints or deadlines: [Review by TPAC 2019 would be appreciated]
  - [X ] I have read and filled out the [Self-Review Questionnare on Security and Privacy](https://www.w3.org/TR/security-privacy-questionnaire/). The [assessment is here](https://w3c.github.io/webrtc-svc/#privacy-security).
  - [X ] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)
  - The group where the work on this specification is: WEBRTC WG

We recommend the explainer to be in [Markdown](https://github.github.com/gfm/). On top of the [usual information expected in the explainer](https://w3ctag.github.io/explainers), it is strongly recommended to add:

  - Links to major pieces of multi-stakeholder review or discussion of this specification: 
  - Links to major unresolved issues or opposition with this specification:

You should also know that...

Relevant background:

Today, multiple browsers have implemented support for Scalable Video Coding (SVC) within their WebRTC implementations.  However, those implementations are proprietary and do not interoperate either at the protocol or API levels.

The goal of the WebRTC-SVC Extension to the WebRTC 1.0 API is to provide an interoperable way for browsers to support configuration of SVC within encoders.   The focus is on the VP8, VP9 and AV1 codecs which support SVC and require that compliant decoders be able to decode anything that an encoder can send.  This simplifies the API because there is no need to negotiate SVC support within Offer/Answer, allowing SVC support to be provided by adding an attribute to the RTCRtpEncodingParameters and RTCRtpCodecCapability dictionaries, without requiring changes to SDP.

The API expresses SVC capabilities and settings by using a shorthand designation for the desired SVC dependency structure known as the "scalability mode".  These scalability modes express commonly utilized configurations of spatial and temporal scalability.  It is also possible to configure spatial simulcast and temporal scalability, either using multiple RTP streams or a single stream (using "S" modes). 

We'd prefer the TAG provide feedback as (please select one):

  - [ ] open issues in our GitHub repo for each point of feedback
  - [X] open a single issue in our GitHub repo for the entire review
  - [ ] leave review feedback as a comment in this issue and @-notify [github usernames]

--------------------------

_Please preview the issue and check that the links work before submitting._ In particular, if anything links to a URL which requires authentication (e.g. Google document), please make sure anyone with the link can access the document.

¹ For background, see our [explanation of how to write a good explainer](https://w3ctag.github.io/explainers).

**Comment by kenchris** [comment]
We usually write mimetypes as all lowercase (the matching of media type and subtype is ALWAYS case-insensitive according to https://tools.ietf.org/html/rfc2045). `video/VP9` looks weird when we are used to write `audio/mp3` etc

**Comment by kenchris** [comment]
Where does the Scalability Mode Identifiers come from? Are these invented by this spec, because if that is the case we usually use lowercase with hyphens and not underscore: https://w3ctag.github.io/design-principles/#casing-rules

**Comment by cynthia** [comment]
This particular example is a bit interesting, in a sense that it assigns to a variable. What is this example attempting to convey? https://w3c.github.io/webrtc-svc/#simulcasttemporal-example*

**Comment by aboba** [comment]
@kenchris The scalability mode identifiers and associated dependency diagrams are defined in the [AV1 bitstream specification](https://aomediacodec.github.io/av1-spec/av1-spec.pdf) Section 6.7.5.

**Comment by aboba** [comment]
@kenchris When used in SDP, H264, VP, VP9 and AV1 are capitalized, as noted in:
[VP8 RTP Payload specification](https://tools.ietf.org/html/rfc7741#page-23),
[VP9 RTP Payload specification](https://tools.ietf.org/html/draft-ietf-payload-vp9-09#page-19),
[H264 RTP Payload specification](https://tools.ietf.org/html/rfc6184#page-58),
[AV1 RTP Payload specification](https://aomediacodec.github.io/av1-rtp-spec/#73-example)

**Comment by aboba** [comment]
@cynthia The variable `encodings` is meant to be used as a parameter in `RTCRtpSender.setParameters()` or `RTCPeerConnection.addTransceiver()` as shown in [Section 10.3 of WebRTC-PC](https://w3c.github.io/webrtc-pc/#simulcast-example).

**Comment by kenchris** [comment]
MIME types are case insensitive. They are lowercase by convention (only).

RFC 2045 says: "The type, subtype, and parameter names are not case sensitive." If you have a MIME type of text/plain that's a type of text and a subtype of plain. So, per the spec, these are not case sensitive.

At least make sure that they work as lowercase as well

**Comment by kenchris** [comment]
A few nits on the spec:

- probably use const (or let) instead of var
- I would prefer there to be no out commented code in examples
- webidl seems to be laid out wrong (too much whitespace)

**Comment by cynthia** [comment]
@aboba Thank you for the clarification. We've looked at it and the intent makes more sense now.

Some questions:
- There are two ways to pass a scalability mode, one in the form of an object, and one in the form of a string. This seems like it can be confusing to the end-user. (meaning developers.)
- Scalability modes are defined as a free form DOMString instead of an enum - any reason for this design decision? It feels like it would make sense to throw in the event an invalid scalability mode is attempted to be used; which is the behavior you would get from an enum.

Otherwise, I think we are happy to see this move forward - and sincere apologies for the huge delay.

**Comment by kenchris** [comment]
I still see comments in the code examples. At least make sure they have the right indentation, but avoid if possible

**Comment by kenchris** [comment]
Ping @aboba

**Comment by aboba** [comment]
@kenchris I have removed the comments, except for Example 1 which is copied from WebRTC-PC Section 7.1, with the exception of a change to the "sendEncodings" dictionary member.

@cynthia This specification doesn't change how encodings are passed in the WebRTC API, it only adds the `scalabilityMode` attribute.  Implementations throw if they encounter values of `scalabilityMode` that they don't support (even if they are valid values).

**Comment by cynthia** [comment]
> Implementations throw if they encounter values of scalabilityMode that they don't support (even if they are valid values).

This is fine, the curiosity was around the choice of a string over an enum, because it seems like the intent is to use a value from a predefined set, which is exactly what a string enum is for.

Is there something in this use case that was not possible with a string enum? (In which case, it might be something that should be addressed in WebIDL.)

**Comment by aboba** [comment]
@cynthia Originally, we referenced the scalability mode names defined in the AV1 specification, but we weren't sure those were the right set of scalability modes, so we used DOMStrings instead of enums. That turned out to be a good choice, because we had to subsequently define our own scalability mode names, which are in some cases different from those used in the AV1 specification (e.g. note the differences in the KEY and KEY_SHIFT modes in the tables).
